core/src/server/method_response.rs

```diff
@@ -0,0 +1,537 @@
+	/// Returns whether the call was successful.
+	pub fn is_success(&self) -> bool {
+		self.success_or_error.is_success()
+	}
+
+	/// Returns whether the call failed.
+	pub fn is_error(&self) -> bool {
+		self.success_or_error.is_success()
+	}
+
+	/// Returns whether the response is a subscription response.
+	pub fn is_subscription(&self) -> bool {
+		matches!(self.kind, ResponseKind::Subscription)
+	}
+
+	/// Returns whether the response is a method response.
+	pub fn is_method_call(&self) -> bool {
+		matches!(self.kind, ResponseKind::MethodCall)
+	}
+
+	/// Returns whether the response is a batch response.
+	pub fn is_batch(&self) -> bool {
+		matches!(self.kind, ResponseKind::Batch)
+	}
+}
+
+/// Represent the outcome of a method call success or failed.
+#[derive(Debug, Copy, Clone)]
+pub enum MethodResponseResult {
+	/// The method call was successful.
+	Success,
+	/// The method call failed with error code.
+	Failed(i32),
+}
+
+impl MethodResponseResult {
+	/// Returns whether the call was successful.
+	pub fn is_success(&self) -> bool {
+		matches!(self, MethodResponseResult::Success)
+	}
+
+	/// Returns whether the call failed.
+	pub fn is_error(&self) -> bool {
+		matches!(self, MethodResponseResult::Failed(_))
+	}
+
+	/// Get the error code
+	///
+	/// Returns `Some(error code)` if the call failed.
+	pub fn as_error_code(&self) -> Option<i32> {
+		match self {
+			Self::Failed(e) => Some(*e),
+			_ => None,
+		}
+	}
+}
+
+impl MethodResponse {
+	/// Consume the method response and extract the serialized response.
+	pub fn into_result(self) -> String {
+		self.result
+	}
+
+	/// Consume the method response and extract the parts.
+	pub fn into_parts(self) -> (String, Option<NotifyKind>) {
+		(self.result, self.on_close)
+	}
+
+	/// Get a reference to the serialized response.
+	pub fn as_result(&self) -> &str {
+		&self.result
+	}
+
+	/// Create a method response from [`BatchResponse`].
+	pub fn from_batch(batch: BatchResponse) -> Self {
+		Self {
+			result: batch.0,
+			success_or_error: MethodResponseResult::Success,
+			kind: ResponseKind::Batch,
+			on_close: None,
+		}
+	}
+
+	/// This is similar to [`MethodResponse::response`] but sets a flag to indicate
+	/// that response is a subscription.
+	pub fn subscription_response<T>(id: Id, result: ResponsePayload<T>, max_response_size: usize) -> Self
+	where
+		T: Serialize + Clone,
+	{
+		let mut rp = Self::response(id, result, max_response_size);
+		rp.kind = ResponseKind::Subscription;
+		rp
+	}
+
+	/// Optional callback that may be utilized to notify that the method response was
+	/// a succesfully JSON-RPC object and processed.
+	pub fn notify_on_success(mut self, tx: MethodResponseNotifyTx) -> Self {
+		self.on_close = Some(NotifyKind::Success(tx));
+		self
+	}
+
+	/// Optional callback that may be utilized to notify that the method response was a
+	/// an JSON-RPC error object and processed.
+	pub fn notify_on_error(mut self, tx: MethodResponseNotifyTx) -> Self {
```

Comment:
Also; how this is set up at the moment means that if somebody does

```rust
response
    .notify_on_success(...)
    .notify_on_error(...)
```

They might expect one of those to fire in each case, but in fact only the last registered one has any chance of ever firing.

I guess I'd either have more than one callback that can be registered, or safer still, just have `notify_on_response` and tell that whether success or error happened.

Reply:
Though I'd also probably want it so that `response.notify_on_completion(a).notify_on_completion(b)` also worked and fired both a and b; one could store a vec of things-to-be-notified perhaps?